lib/pgBackRest/Storage/S3/Driver.pm

```diff
@@ -0,0 +1,512 @@
+        );
+
+    # Determine the prefix (this is the search path within the bucket)
+    my $strPrefix = $strPath eq qw{/} ? undef : substr($strPath, 1) . ($bPath ? qw{/} : '');
+
+    # A delimiter must be used if recursion is not desired
+    my $strDelimiter = $bRecurse ? undef : '/';
+
+    # Hash to hold the manifest
+    my $hManifest = {};
+
+    # Continuation token - returned from requests where there is more data to be fetched
+    my $strContinuationToken;
+
+    my $iFileTotal = 0;
+    my $iPathTotal = 0;
+
+    do
+    {
+        # Get the file list
+        my $oResponse = $self->request(
+            HTTP_VERB_GET, {hQuery =>
+            {&S3_QUERY_LIST_TYPE => 2, &S3_QUERY_PREFIX => $strPrefix, &S3_QUERY_DELIMITER => $strDelimiter,
+                &S3_QUERY_CONTINUATION_TOKEN => $strContinuationToken}, strResponseType => S3_RESPONSE_TYPE_XML});
+
+        # Modify the prefix for file searches so the filename is not stripped off
+        if (defined($strPrefix) && !$bPath)
+        {
+            # If there are no paths in the prefix then undef it
+            if (index($strPrefix, qw{/}) == -1)
+            {
+                undef($strPrefix);
+            }
+            else
+            {
+                $strPrefix = dirname($strPrefix) . qw{/};
+            }
+        }
+
+        # Store files
+        foreach my $oFile (xmlTagChildren($oResponse, "Contents"))
+        {
+            my $strName = xmlTagText($oFile, "Key");
+
+            # Strip off prefix
+            if (defined($strPrefix))
+            {
+                $strName = substr($strName, length($strPrefix));
+            }
+
+            $hManifest->{$strName}->{type} = 'f';
+            $hManifest->{$strName}->{size} = xmlTagText($oFile, "Size");
+            $iFileTotal++;
+
+            # Generate paths from the name if recursing
+            if ($bRecurse)
+            {
+                my @stryName = split(qw{/}, $strName);
+
+                if (@stryName > 1)
+                {
+                    $strName = undef;
+
+                    for (my $iIndex = 0; $iIndex < @stryName - 1; $iIndex++)
+                    {
+                        $strName .= (defined($strName) ? qw{/} : '') . $stryName[$iIndex];
+                        $hManifest->{$strName}->{type} = 'd';
+                    }
+                }
+            }
+        }
+
+        # Store directories
+        if ($bPath && !$bRecurse)
+        {
+            foreach my $oPath (xmlTagChildren($oResponse, "CommonPrefixes"))
+            {
+                my $strName = xmlTagText($oPath, "Prefix");
+
+                # Strip off prefix
+                if (defined($strPrefix))
+                {
+                    $strName = substr($strName, length($strPrefix));
+                }
+
+                # Strip off final /
+                $strName = substr($strName, 0, length($strName) - 1);
+
+                $hManifest->{$strName}->{type} = 'd';
+                $iPathTotal++;
```

Comment:
What is $iPathTotal ever actually used for? Was it intended to be part of the while condition?